qp_spades/qp_spades.py

```diff
@@ -0,0 +1,153 @@
+
+    # Step 2 generating command
+    qclient.update_job_step(job_id, "Step 2 of 4: Generating commands")
+
+    flash_cmd = (f'flash --threads {parameters["threads"]} --max-overlap=%d '
+                 f'--max-mismatch-density=0.1 --output-directory {out_dir} '
+                 '--output-prefix="%s" %s %s')
+    spades_cmd = (
+        f'spades.py --{parameters["type"]} -t {parameters["threads"]} '
+        f'-m {parameters["memory"]} -k {parameters["k-mers"]} -1 %s -2 %s '
+        f'-o {out_dir}/%s')
+
+    commands = []
+    outfiles = []
+    overlap = None
+
+    for fwd_fp, rev_fp in zip(fwd_seqs, rev_seqs):
+        fwd_fn = basename(fwd_fp)
+
+        run_prefix = None
+        for rp in prefix_to_name:
+            if fwd_fn.startswith(rp):
+                if run_prefix is None:
+                    run_prefix = rp
+                else:
+                    msg = ('Multiple run prefixes match this fwd '
+                           'file: {fwd_fn}')
+                    return False, None, msg
+
+        rev_fn = basename(rev_fp)
+        # if we have reverse reads, make sure the matching pair also
+        # matches the run prefix:
+        if not rev_fn.startswith(run_prefix):
+            msg = ('Reverse read does not match this run prefix.\nRun prefix: '
+                   '%s\nForward read: %s\nReverse read: %s\n' % (
+                    run_prefix, fwd_fn, rev_fn))
+            return False, None, msg
+
+        sample_name = prefix_to_name[run_prefix]
+        out_fp = f'{out_dir}/{sample_name}.fasta'
+        if parameters['merging'] == 'no merge':
+            cmd = (f'{spades_cmd} > {out_dir}/%s.spades.log 2>&1; '
+                   f'mv {out_dir}/{sample_name}/scaffolds.fasta {out_fp}' % (
+                    fwd_fp, rev_fp, sample_name, sample_name))
+        elif parameters['merging'].startswith('flash '):
+            if overlap is None:
+                # get read lenght quickly; note that we are going to assume
+                # that (1) the forward and reverse are the same lenght and (2)
+                # all file pairs have the same lenght so only calculate once
+                std_out, std_err, return_value = system_call(
+                    f'gunzip -c {fwd_fp} | head -n 2')
+                if return_value != 0:
+                    error_msg = (f"Error uncompressing: {fwd_fp}\n"
+                                 f"Std out: {std_out}\nStd err: {std_err}\n")
+                    return False, None, error_msg
+
+                read_length = len(std_out.split('\n')[1])
+                percentage = int(parameters['merging'][6:-1])/100
+                overlap = int(read_length * percentage)
+
+            flash_fwd = f'{out_dir}/{sample_name}.notCombined_1.fastq'
+            flash_rev = f'{out_dir}/{sample_name}.notCombined_2.fastq'
+            flash_merge = f'{out_dir}/{sample_name}.extendedFrags.fastq'
+            cmd = (f'{flash_cmd} > {out_dir}/%s.flash.log 2>&1; '
```

Comment:
Same as before, if there's a failure anywhere in the pipeline we'll probably see an uninformative message. Maybe break these down or separate them by `&&`.